packages/@react-aria/autocomplete/src/useAutocomplete.ts

```diff
@@ -0,0 +1,146 @@
+ * @param state - State for the autocomplete, as returned by `useAutocompleteState`.
+ */
+export function useAutocomplete<T>(props: AriaAutocompleteOptions, state: AutocompleteState): AutocompleteAria<T> {
+  let {
+    inputRef,
+    isReadOnly
+  } = props;
+
+  let menuId = useId();
+
+  // TODO: may need to move this into Autocomplete? Kinda odd to return this from the hook? Maybe the callback should be considered
+  // external to the hook, and the onus is on the user to pass in a onKeydown to this hook that updates state.focusedNodeId in response to a key
+  // stroke
+  let callbackRef = useRef<(e: KeyboardEvent) => string>(null);
+  let register = (callback) => {
+    callbackRef.current = callback;
+  };
+
+  // For textfield specific keydown operations
+  let onKeyDown = (e: BaseEvent<KeyboardEvent<any>>) => {
+    if (e.nativeEvent.isComposing) {
+      return;
+    }
+
+    // TODO: how best to trigger the focused element's action? Currently having the registered callback handle dispatching a
+    // keyboard event
+    switch (e.key) {
+      case 'Escape':
+        if (state.inputValue !== '' && !isReadOnly) {
+          state.setInputValue('');
+        } else {
+          e.continuePropagation();
+        }
+
+        break;
+      case 'Home':
+      case 'End':
+      case 'ArrowUp':
+      case 'ArrowDown':
+        // Prevent these keys from moving the text cursor in the input
+        e.preventDefault();
+        break;
+    }
+
+    if (callbackRef.current) {
+      let focusedNodeId = callbackRef.current(e);
+      state.setFocusedNodeId(focusedNodeId);
+    }
+  };
+
+  let {labelProps, inputProps, descriptionProps} = useTextField({
+    ...props as any,
+    onChange: state.setInputValue,
+    onKeyDown: chain(onKeyDown, props.onKeyDown),
+    value: state.inputValue,
+    autoComplete: 'off'
+  }, inputRef);
+
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-aria/autocomplete');
+  let menuProps = useLabels({
+    id: menuId,
+    'aria-label': stringFormatter.format('menuLabel'),
+    'aria-labelledby': props['aria-labelledby'] || labelProps.id
+  });
+
+  return {
+    labelProps,
+    inputProps: mergeProps(inputProps, {
+      'aria-haspopup': 'listbox',
+      'aria-controls': menuId,
+      // TODO: readd proper logic for completionMode = complete (aria-autocomplete: both)
+      'aria-autocomplete': 'list',
+      'aria-activedescendant': state.focusedNodeId ?? undefined,
+      // TODO: note that the searchbox role causes newly typed letters to interrupt the announcement of the number of available options in Safari.
+      // I tested on iPad/Android/etc and the combobox role doesn't seem to do that but it will announce that there is a listbox which isn't true
+      // and it will say press Control Option Space to display a list of options which is also incorrect. To be fair though, our combobox doesn't open with
+      // that combination of buttons
+      role: 'searchbox',
```

Comment:
something to note with the announcements, hopefully this won't be a problem as browsers/screenreaders begin to handle announcing of aria-activedescendant properly